File: binsrc/genie/Configuration.cs
using System;
using System.Collections.Generic;
using System.IO;
using Microsoft.Extensions.Configuration;

namespace ArkaneSystems.WindowsSubsystemForLinux.Genie
{
    /// <summary>
    /// Configuration values for Genie.
    /// </summary>
    internal class GenieConfig
    {
        #region Static configuration

        // Install location of genie and friends.
#if LOCAL
        internal const string Prefix = "/usr/local";
#else
        private const string Prefix = "/usr";
#endif

        /// <summary>
        /// Default environment variables to be added to every genie bottle.
        /// </summary>
        internal static readonly string[] DefaultVariables = { "INSIDE_GENIE=true" };

        #endregion Static configuration

        private IConfiguration Configuration { get; }

        internal GenieConfig()
        {
            Configuration = new ConfigurationBuilder().SetBasePath("/etc").AddIniFile("genie.ini", optional: false).Build();
        }

        /// <summary>
        /// Secure path to use when shelling out from Genie and to init systemd.
        /// </summary>
        internal string SecurePath => Configuration.GetValue("genie:secure-path", "/lib/systemd:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin");

        /// <summary>
        /// True if the original path should be merged onto the end of the secure path for systemd and/or genie bottle;
        /// false otherwise.
        /// </summary>
        internal bool ClonePath => Configuration.GetValue("genie:clone-path", false);

        /// <summary>
        /// Environment variables to copy into the systemd environment and/or genie bottle.
        /// </summary>
        internal IEnumerable<string> CloneEnv => Configuration.GetValue("genie:clone-env", "WSL_DISTRO_NAME,WSL_INTEROP,WSLENV")
            .Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);

        /// <summary>
        /// True to update the host name with the "-wsl" suffix, false otherwise.
        /// </summary>
        internal bool UpdateHostname => Configuration.GetValue("genie:update-hostname", true);

        /// <summary>
        /// Suffix with which to update the host name, if the above is true.
        /// </summary>
        internal string HostnameSuffix => Configuration.GetValue("genie:update-hostname-suffix", "-wsl");

        /// <summary>
        /// Path to the local binary for unshare(1).
        /// </summary>
        internal string PathToUnshare => Configuration.GetValue("genie:unshare", "/usr/bin/unshare");

        /// <summary>
        /// Seconds to wait for systemd to enter the running state on startup.
        /// </summary>
        internal int SystemdStartupTimeout => Configuration.GetValue("genie:systemd-timeout", 240);

        /// <summary>
        /// True to symlink a stub resolv.conf file for systemd-resolved, false otherwise.
        /// </summary>
        internal bool ResolvedStub => Configuration.GetValue("genie:resolved-stub", false);

        internal bool AppArmorNamespace => Configuration.GetValue("genie:apparmor-namespace", false);

        /// <summary>
        /// Get the installation-dependent path for a given file.
        /// </summary>
        internal static string GetPrefixedPath(string path) => Path.Combine(Prefix, path);

        /// <summary>
        /// Warn the user if the systemd target is not multi-user.target.
        /// </summary>
        internal bool TargetWarning => Configuration.GetValue("genie:target-warning", true);
    }
}

File: binsrc/genie/configuration.py
# Configuration data module

import configparser

# Global variables

_config = None


# functions
def clonable_envars():
    """Get the list of environment variables to clone."""
    return (_config.get('genie', 'clone-env',
                        fallback='WSL_DISTRO_NAME,WSL_INTEROP,WSLENV,DISPLAY,WAYLAND_DISPLAY,PULSE_SERVER')).split(',')


def clone_path():
    """Do we clone the outside-bottle path, or not?"""
    return _config.getboolean('genie', 'clone-path', fallback=False)


def resolved_stub():
    """Do we make the systemd-resolved stub, or not?"""
    return _config.getboolean('genie', 'resolved-stub', fallback=False)


def secure_path():
    """Get the configured secure path."""
    return _config.get('genie', 'secure-path', fallback='/lib/systemd:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin')


def system_timeout():
    """Return the configured timeout for systemd startup."""
    return _config.getint('genie', 'systemd-timeout', fallback=240)


def target_warning():
    """Warn when the systemd target is not set to 'multi-user.target'."""
    return _config.getboolean('genie', 'target-warning', fallback=True)


def update_hostname():
    """Update the hostname in the config file?"""
    return _config.getboolean('genie', 'update-hostname', fallback=True)


def update_hostname_suffix():
    """Update the hostname suffix in the config file with..."""
    return _config.get('genie', 'update-hostname-suffix', fallback='-wsl')


# Initialization

def load():
    """Load the configuration from the config file ('/etc/genie.ini')."""
    global _config

    _config = configparser.ConfigParser()
    _config.read('/etc/genie.ini')
